File: golang/q400/q441.go
package q400

import "math"

func ArrangeCoins(n int) int {
	// total = (1 + n) * n / 2
	// total * 2 = n + n^2
	// n^2 + n - total * 2 = 0
	// n = (-1 + sqrt(1 + 8 * total))/2
	x := (-1.0 + math.Sqrt(1.0+8.0*float64(n))) / 2
	return int(x)
}

File: hs/src/Q400/Q441.hs
module Q400.Q441 
    (arrangeCoins,
     arrangeCoinsPlanA
    ) where

arrangeCoins :: Integer -> Integer
arrangeCoins 0 = 0
arrangeCoins n = floor $ ((sqrt . fromIntegral $ (1+8*n)::Double) - 1) / 2


arrangeCoinsPlanA :: Integer -> Integer
arrangeCoinsPlanA 0 = 0
arrangeCoinsPlanA n = pa 1 n
        where
            pa num total
                | total >= num = pa (num+1) (total-num)
                | otherwise = num -1
